src/table.rs

```diff
@@ -165,6 +169,113 @@ where
+        let end = start
+            + binary_search(&self.table[start..], |&sufi| {
+                !self.text[sufi as usize..].starts_with(query)
+            });
+
+        (start, end)
+    }
+
+    /// Returns an unordered list of positions where `query` starts in `text`, limiting the search to a
+    /// specified range of the suffix table.
+    fn range_positions(&self, query: &[u16], range_start: usize, range_end: usize) -> &[u64] {
+        if self.text.len() == 0
+            || query.len() == 0
+            || (query < self.suffix(0 + range_start) && !self.suffix(0 + range_start).starts_with(query))
+            || query > self.suffix(std::cmp::max(0, range_end - 1))
+        {
+            return &[];
+        }
+
+        let start = binary_search(&self.table[range_start..range_end], |&sufi| query <= &self.text[sufi as usize..]);
+        let end = start
+            + binary_search(&self.table[range_start + start..range_end], |&sufi| {
+                !self.text[sufi as usize..].starts_with(query)
+            });
+    
+        if start > end {
+            &[]
+        } else {
+            &self.table[range_start + start..range_start + end]
+        }
+    }
+
+    /// Returns an unordered list of counts of token values that succeed `query`.
+    /// Counts all tokens if query is empty.
+    fn next_token_counts(&self, query: &[u16], vocab: Option<u16>) -> Vec<usize> {
+        let mut counts: Vec<usize> = vec![0usize; vocab.unwrap_or(u16::MAX) as usize + 1];
+        let mut suffixed_query = query.to_vec();
+        let (range_start, range_end) = self.boundaries(query);
+
+        for i in 0..counts.len() {
+            suffixed_query.push(i as u16);
+            
+            let positions = self.range_positions(&suffixed_query, range_start, range_end);
+            counts[i] = positions.len();
+            suffixed_query.pop();
+        }
+        counts
+    }
+
+    /// Count the occurrences of each token that directly follows each query sequence.
+    pub fn batch_next_token_counts(&self, queries: &[Vec<u16>], vocab: Option<u16>) -> Vec<Vec<usize>> {
+        queries.into_par_iter()
+            .map(|query| {
+                self.next_token_counts(query, vocab)
+            })
+            .collect()
+    }
+
+    /// Autoregressively sample k characters from a conditional distribution based 
+    /// on the previous (n - 1) characters (n-gram prefix) in the sequence.
+    pub fn sample(&self, query: &[u16], n: usize, k: usize) -> Result<Vec<u16>> {
+        let mut rng = thread_rng();
+        let mut sequence = Vec::from(query);
+
+        for _ in 0..k {
+            // look at the previous (n - 1) characters to predict the n-gram completion
+            let start = sequence.len().saturating_sub(n as usize - 1);
+            let prev = &sequence[start..];
+            
+            let counts: Vec<usize> = self.next_token_counts(prev, Option::None);
```

Comment:
nit: does this actually need to be `Option::None`? I thought `None` was always in the public namespace or whatever

Reply:
Oh yeah it is in the namespace cheers!